### DISCUSSION TITLE
Working OAuth example?

### DISCUSSION THREAD

**Discussion by marckohlbrugge** [bluesky-social/atproto#3075] [How to do it?] [answered]
For the last week I've tried to implement ATProto OAuth with Ruby, but it's proven to be very complicated. Although there are numerous existing OAuth libraries, none implement the more modern (?) feature set that AT Proto requires, such as DPoP.

Finally haven given up on implementing it in Ruby, I decided to try Node instead as there's some official documentation and libraries available. Unfortunately, I've not succeeded to implement OAuth here either.

Although there are quite a few code examples available, none seem complete.

For example, there this:

https://www.npmjs.com/package/@atproto/oauth-client-node

Here's some of the issues with it I've found so far:

- It uses `'https://my-app.com/callback'` as an example, but the Express route is `/atproto-oauth-callback`
- There is no explanation on how to generate a working private key (I finally was able to do this after much trial and error)
- It uses `new Agent(session)` yet `Agent` is not defined
- When trying to import `Agent`, there's a message saying it's not a constructor
- `stateStore` and `sessionStore` are not implemented (perhaps easy for experienced Node developers, but for someone with a different background like me this is yet another hurdle to overcome)
- `requestLock` is not explained

There's probably more issues, but I got stuck with the `Agent` not being authorized for some reason.

AT Proto looks promising, but without complete code examples it's very difficult for developers to integrate with it.

I'd like to propose for someone to create a complete (yet basic) code example of `@atproto/oauth-client-node` that works and is up-to-date.

**Comment by sh-k-l** [comment]
Here's a working example I found: https://github.com/pilcrowonpaper/atproto-oauth-example/tree/main

**Reply by apsquared** [reply]
Would this approach allow a user to login with bluesky and then allow my app to post on that user's behalf?

**Reply by jay-elizaga-dev** [reply]
Be careful copying the private key contents that the readme shows publicly: https://github.com/pilcrowonpaper/atproto-oauth-example/tree/main?tab=readme-ov-file#deploying-to-production

Please do NOT publicly release private keys.

**Comment by majkelcc** [comment]
Hi @marckohlbrugge, did you manage to make your code work somehow?

**Comment by brenc** [comment]
I'd love to see a complete working example using Typescript too (Bun, Node, Deno) for a bot account. Having a hard time figuring out how to switch to OAuth session management as the docs recommend doing.

**Comment by kidGodzilla** [comment]
I put together a node/express vanilla js example, if it's helpful to anyone who finds this in the future.

~~My last Todo (I think) is JWKS:~~

https://github.com/kidGodzilla/bsky-oauth-example

~~@apsquared you might need token storage and refresh as well if you want posts to happen while the user is not signed into your app.~~ I've verified the node library handles token storage and refresh as long as you persist your session store or it survives the refresh cycle.

Edit: I also added a PR for what I learned during implementation that I wished was in the README.md:

https://github.com/bluesky-social/atproto/pull/3254

**Comment by bnewbold** [comment, marked as answer]
I recommend looking at this OAuth announcement blog post: https://docs.bsky.app/blog/oauth-atproto

It links out to several resources, including an "OAuth Client Implementation Guide", and working client code in Python (which uses very few dependencies).

**Comment by alex97po** [comment]
Hello, everyone!
I am working on a social media planner app and to have UX similar to all other platforms I would like user to enroll their Bluesky account using OAuth flow. 
I constructed `https://bsky.social/oauth/authorize` url:
`window.location.href = https://bsky.social/oauth/authorize?response_type=code&client_id=${METADATA_JSON_URL}&redirect_uri=${encodeURIComponent(redirectUri)}&scope=atproto&state=${state}&code_challenge=${codeChallenge}&code_challenge_method=S256;`
In callback I receive iss, state, and code - I check state and then pass code and code verifier to backend for token exchange. 
Backend calls 
```
curl --location 'https://bsky.social/oauth/token' \
--header 'Content-Type: application/x-www-form-urlencoded' \
--data-urlencode 'code=cod-561b523bf82ae06bb3adc6bca07cd6488002a3e9d410e4c032629c3d44833877' \
--data-urlencode 'grant_type=authorization_code' \
--data-urlencode 'code_verifier=AArJDWAkJHB5W1edYUFBKeLHKL9RuVPS.aPRz-9NdOKlMXLjMWRkY7yEGmwPzUajEeiFLn1hv0LBdzrB3pni~i~yRSnX~bCSv.X_I8kx2wcw' \
--data-urlencode 'redirect_uri=same redirect uri as was used on step 1' \
--data-urlencode 'client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-bearer' \
--data-urlencode 'client_assertion=token' \
--data-urlencode 'client_id=https://post-pulse.com/oauth/client-metadata.json'
```
I fixed few errors - I was missing few claims in my JWT initially, my public key JWK was invalid - all is fixed. 
But I am facing `{"error":"invalid_request","error_description":"Invalid code"}`.
100% I am using code one time only and no, there is no delay between frontend and backend. 
I refactored flow with PAR although it is different from other platforms flow - FE first calls BE for PAR url - generated on backend with client_id, request, client_assertion_type, client_assertion - same problem - I am redirected to Bluesky login, login, redirected back with code - call backend, backend calls token exchange API - same invalid token error.
May be I am wrong, but since authorization works well and I am facing problem with token exchange API only, I suspect that my client is treated as public and not confidential. 
On the other hand, I do have jwks_uri in my metadata json - I Bluesky did recognize it (because I faced some errors related to my JWK, as I mentioned). 
What am I missing?

**Comment by mickaelvieira** [comment]
If anyone is looking for an example in Go, I wrote [a small service](https://github.com/mickaelvieira/atproto-oauth2-go-example) demonstrating the full authorization flow (Access Token Request & Refresh Token Request) based on this [Go example](https://github.com/potproject/atproto-oauth2-go-example) & the [Python example](https://github.com/bluesky-social/cookbook/tree/main/python-oauth-web-app) in the cookbook repository. I thought this might help someone implementing their version.

**Comment by AndreasThinks** [comment]
Just to +1 this one... the fact there is a working oauth example flow in Python flow anywhere (the version on the cookbook specifically states this is "hard mode" rather than using the sdk) is a bit sad. Its easy enough using login + app password, but an easy oauth implementation into the SDK would be very appreciated!

**Reply by alex97po** [reply]
We eventually managed to implement Bluesky OAuth, although it proved to be much more complicated than integrating with other social media platforms. Perhaps my article on Medium will be of some help - https://medium.com/p/f215cdbcadda. Code samples are in Java, but I am sure it won't be hard to translate them to Python.
In a nutshell, the process is the following:
1. Usually, the frontend constructs an authorization call and redirects the user to the corresponding platform. Like, https://api.instagram.com/oauth/authorize?client_id=... For Bluesky, the authorize URL requires a client assertion JWT token, so you need to construct it on the backend with a private key (check the code sample in the article).
2. User completes authorization and is redirected back to your app. The app receives the code_verifier and other typical OAuth parameters and passes them to the backend. Nothing special here.
3. Backend calls the token API for code exchange. Here, the trickiest part comes into play, which made me nuts - DPoP. The backend needs to generate a DPoP JWT without the nonce field (see the code sample in the article). Call `https://bsky.social/oauth/token` with the DPoP header (PAY ATTENTION, HEADER SHOULD BE CALLED DPoP NOT AUTHORIZATION, NO BEARER PREFIX, JUST JWT HERE) and accept the error response that will specify the use_dpop_nonce. Then you take DPoP-Nonce response header value and start over - generate DPoP one more time (THIS TIME WITH NONCE), generate client assertion again (to be included in form data together with other OAuth fields), and make a call.
This trick with DPoP and "double call" applies to other endpoints as well, e.g., `/xrpc/com.atproto.repo.createRecord `

**Reply by AndreasThinks** [reply]
I had a very similar adventure yesterday! Nearly got there, but then it broke my use of the Client in the SDK so I sort of gave up...
